README.md
# bitcoinnode

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2019 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-01-14
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-01-14
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/bitcoinnode/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/bitcoinnode/compare/0.1.0...0.1.1


test/bitcoinnode/core_test.clj
(ns bitcoinnode.core-test
  (:require [clojure.test :refer :all]
            [bitcoinnode.core :refer :all])
  (:import [java.io ByteArrayInputStream]))


(def valid-payload    
  [0x62 0xEA 0x00 0x00                                                                   ; 60002 (protocol version 60002)
   0x01 00 00 00 00 00 00 00                                                       ; 1 (NODE_NETWORK services)
   0x11 0xB2 0xD0 0x50 00 00 00 00                                                       ; Tue Dec 18 10:12:33 PST 2012
   0x01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0xFF 0xFF 00 00 00 00 00 00 ; Recipient address info - see Network Address
   0x01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 0xFF 0xFF 00 00 00 00 00 00 ; Sender address info - see Network Address
   0x3B 0x2E 0xB3 0x5D 0x8C 0xE6 0x17 0x65                                                       ; Node ID
   0x0F 0x2F 0x53 0x61 0x74 0x6F 0x73 0x68 0x69 0x3A 0x30 0x2E 0x37 0x2E 0x32 0x2F                               ; "/Satoshi:0.7.2/" sub-version string (string is 15 bytes long)
   0xC0 0x3E 0x03 0x00])

(def valid-ver-msg 
  (concat
    [0xF9 0xBE 0xB4 0xD9                                                                   ;Main network magic bytes
     0x76 0x65 0x72 0x73 0x69 0x6F 0x6E 0x00 0x00 0x00 0x00 0x00                                           ;"version" command
     0x64 0x00 0x00 0x00                                                                   ;Payload is 100 bytes long
     0x35 0x8d 0x49 0x32]
    valid-payload))

["f9" "be" "b4" "d9" 
 "76" "65" "72" "73" "69" "6f" "6e" "0" "0" "0" "0" "0" 
 "66" "0" "0" "0" 
 "ed" "b3" "e5" "18" 
 "7f" "11" "1" "0" 
 "d" "4" "0" "0" "0" "0" "0" "0" 
 "89" "85" "41" "5c" "0" "0" "0" "0" 
 "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "ff" "ff" "d4" "10" "b0" "1a" "16" "3a" 
 "d" "4" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "94" "94" "e0" "96" "8e" "b3" "54" "a6" "10" "2f" "53" "61" "74" "6f" "73" "68" "69" "3a" "30" "2e" "31" "37" "2e" "30" "2f" "99" "87" "8" "0" "1" "f9" "be" "b4" "d9" "76" "65" "72" "61" "63" "6b" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "5d" "f6" "e0" "e2" "f9" "be" "b4" "d9" "61" "6c" "65" "72" "74" "0" "0" "0" "0" "0" "0" "0" "a8" "0" "0" "0" "1b" "f9" "aa" "ea" "60" "1" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "0" "ff" "ff" "ff" "7f" "0" "0" "0" "0" "ff" "ff" "ff" "7f" "fe"]

(def real-reply-from-39-106-249-60 [249 190 180 217 118 101 114 115 105 111 110 0 0 0 0 0 102 0 0 0 237 179 229 24 127 17 1 0 13 4 0 0 0 0 0 0 137 133 65 92 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 255 255 212 16 176 26 22 58 13 4 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 148 148 224 150 142 179 84 166 16 47 83 97 116 111 115 104 105 58 48 46 49 55 46 48 47 153 135 8 0 1 249 190 180 217 118 101 114 97 99 107 0 0 0 0 0 0 0 0 0 0 93 246 224 226 249 190 180 217 97 108 101 114 116 0 0 0 0 0 0 0 168 0 0 0 27 249 170 234 96 1 0 0 0 0 0 0 0 0 0 0 0 255 255 255 127 0 0 0 0 255 255 255 127 254])
(deftest decode-roundtrip 
  (is (= valid-ver-msg (msg->bytes (read-msg (ByteArrayInputStream. (byte-array valid-ver-msg)) mainnet) mainnet))))



 


src/bitcoinnode/core.clj
(ns bitcoinnode.core
  (:use [clojure.pprint])
  (:import [java.net Socket]
           [java.io BufferedInputStream BufferedOutputStream ByteArrayInputStream]
           [java.nio.charset StandardCharsets]
           [java.time Instant]
           [java.nio.charset StandardCharsets]
           [org.apache.commons.codec.binary Hex]
           [java.security MessageDigest]))
(defmacro dbg [body]
  `(let [x# ~body]
     (println "dbg:" '~body "=" x#)
     x#))

(defn pad [n coll val]
  (take n (concat coll (repeat val))))

(defn endian-of [r endian]
  (condp = endian
	  :little
	  r
	  :big
	  (reverse r)))


(defn bytes->int [bytes endian]
    (reduce (fn [acc v] (bit-or (bit-shift-left acc 8) v)) (endian-of bytes endian)))

(defn invert-endian [endian]
  (condp = endian
    :little :big
    :big :little))

(defn int->bytes [value nr-of-bytes endian]
    (map #(bit-and (bit-shift-right value (* % 8)) 0xff) (endian-of (range nr-of-bytes) (invert-endian endian))))

(def sha-256 (MessageDigest/getInstance "SHA-256"))

(defn bytes->sha256 [bytes]
    (.digest sha-256 bytes))

(defn bytes->unsigned [bytes]
  (map #(bit-and % 0xff) bytes))


(defn calc-checksum-as-bytes [bytes]
  (let [cs (bytes->unsigned (->> bytes bytes->sha256 bytes->sha256 (take 4)))]
    (map #(Integer/toString % 16) cs)
    cs))

(defn calc-checksum [bytes]
  (bytes->int (calc-checksum-as-bytes bytes) :little))
  



(defn read-bytes [in n] 
  (let [bytes (byte-array n)]
    (.read in bytes)
    (bytes->unsigned bytes)))

(defn read-int 
  ([in nr-of-bytes endian]
  (bytes->int (read-bytes in nr-of-bytes) endian))
  ([in nr-of-bytes]
    (read-int in nr-of-bytes :little)))

(defn read-bool [in]
  (read-int in 1))

(defn read-var-int [in]
  (let [first-byte (read-int in 1)]
    (cond 
      (= first-byte 0xff) (read-int in 8)
      (= first-byte 0xfe) (read-int in 4)
      (= first-byte 0xfd) (read-int in 2)
      :else
      first-byte)))
  
(defn read-var-str [in]
  (let [n (read-var-int in)]
    (String. (byte-array (read-bytes in n)) StandardCharsets/US_ASCII)))

(defn str->var-bytes [s]
  (let [n (count s)
        bytes (.getBytes s)]
    (cond 
      (< n 0xfd) (concat [n] bytes)
      (<= n Short/MAX_VALUE) (concat [0xfd] (int->bytes n 2 :little) bytes) 
      (<= n Short/MAX_VALUE) (concat [0xfe] (int->bytes n 4 :little) bytes)
      :else (concat [0xff] (int->bytes n 8 :little) bytes)
    )))

(defn read-timestamp [in nr-of-bytes]
  (read-int in nr-of-bytes))


(defn read-magic [in]
  (read-int in 4 :big))

(defn read-cmd [in]
  (String. (byte-array (take-while #(> % 0) (read-bytes in 12))) StandardCharsets/US_ASCII))
  
(defn cmd->bytes [cmd]
  (pad 12 (.getBytes cmd) 0))
  

(defn read-length [in]
  (read-int in 4 :big))

(defn read-checksum [in]
  (read-int in 4))

(defn read-payload [in n checksum]
  (let [pl (byte-array (read-bytes in n))]
    ;(ByteArrayInputStream. (byte-array pl))
    (if (= (calc-checksum pl) checksum)
      (ByteArrayInputStream. pl)
      (throw (IllegalStateException. "Wrong checksum in payload")))))


(defn int->services [v]
  {:node-network (= (bit-and v 0x1) 0x1)
     :node-get-utxo (= (bit-and v 0x2) 0x2)
     :node-bloom (= (bit-and v 0x4) 0x4)
     :node-witness (= (bit-and v 0x8) 0x8)
     :node-network-limited (= (bit-and v 1024) 1024)})

(defn services->int [services]
  (bit-or
    (if (:node-network services) 1 0)
    (if (:node-get-utxo services) 2 0)
    (if (:node-bloom services) 4 0)
    (if (:node-witness services) 8 0)
    (if (:node-network-limited services) 1024 0)
    ))

(defn read-services [in]
  (int->services (read-int in 8 :big)))


;(defn bytes->ip [bytes]

(defn read-ip [in]
  (doseq [b (read-bytes in 10)]
    (when (not= b 0) (throw (IllegalStateException. (format "byte should be 0 but is %s" b)))))
  (read-bytes in 2)
  #_(doseq [b (read-bytes in 2)]
     (when (not= b 255) (throw (IllegalStateException. (format "byte should be 255 but is %s" b)))))
   (dbg (read-bytes in 4)))

(defn ip->bytes [ip]
  (concat
    [0 0 0 0 0 0 0 0 0 0 0xff 0xff]
    ip))

(defn read-port [in]
  (read-int in 2 :big))

(defn read-network-address 
  ([in version-cmd?]
  ;TODO time  
  {:time (when (not version-cmd?) (read-timestamp in 4))
   :services (read-services in)
   :ip (read-ip in)
   :port (read-port in)
   }
  )
  ([in]
    (read-network-address in false)))

(defn network-address->bytes 
  ([addr version-cmd?]
  ;TODO time  
  (concat
    (if version-cmd? [] (int->bytes (:time addr) 4 :little))
    (int->bytes (services->int (:services addr)) 8 :big)
    (ip->bytes (:ip addr))
    (int->bytes (:port addr) 2 :big)
    )
  )
  ([addr]
    (network-address->bytes addr false)))
    
                              

(defn read-version [in]
  (read-int in 4 :big))

(defmulti decode-cmd (fn [cmd in] cmd))
(defmulti encode-cmd (fn [cmd payload] cmd))


(defn read-msg [in magic]
  (let [m (read-magic in)]
    (if (= m magic)
      (let [m {:magic m
               :cmd (read-cmd in)
               :length (read-length in)
               :checksum (read-checksum in)
               }
            ]
        (assoc m :payload (decode-cmd (:cmd m) (read-payload in (:length m) (:checksum m)))))
      (throw (IllegalStateException. "Wrong network magic value")))))
  

(defn msg->bytes [msg magic]
  (let [payload (encode-cmd (:cmd msg) (:payload msg))]
    (concat 
      (int->bytes magic 4 :big)
      (cmd->bytes (:cmd msg))
      (int->bytes (count payload) 4 :big)
      (calc-checksum-as-bytes (byte-array payload))
      payload
      ))) 


(defn write-msg! [data out]
  (.write out (byte-array data)))
  
(defn decode-version-cmd>=70001 [version cmd in]
  (if (>= version 70001)
    (assoc 
      cmd :relay (if (= (read-int in 1) 0) false true))
    cmd))
  

(defn decode-version-cmd>=106 [version cmd in]
  (if (>= version 106)
    (decode-version-cmd>=70001 version
      (assoc cmd
        :addr-from (read-network-address in true)
        :nounce (read-int in 8 :big) 
        :user-agent (read-var-str in)
        :start-height (read-int in 4 :big))
      in)
    cmd))

(defmethod decode-cmd "version" [_ in] 
  (let [cmd {:version (read-version in)
             :services (read-services in)
             :timestamp (read-timestamp in 8)
             :addr-recv (read-network-address in true)
             }
     version (:version cmd)
     ]
    (decode-version-cmd>=106 version cmd in)))
    


(defn encode-version-cmd>=106 [version pl]
  (if (>= version 106)
    (concat
      (network-address->bytes (:addr-from pl) true)
      (int->bytes (:nounce pl) 8 :big)
      (str->var-bytes (:user-agent pl))
      (int->bytes (:start-height pl) 4 :big)
      )
    []))

(defn encode-version-cmd>=70001 [version pl]
  (if (>= version 70001)
   (int->bytes (:relay pl) 1 :little)
    []))

(defmethod encode-cmd "version" [_ payload]
  (let [version (:version payload)]
    (concat
      (int->bytes version 4 :big) 
      (int->bytes (services->int (:services payload)) 8 :big)
      (int->bytes (:timestamp payload) 8 :little)
      (network-address->bytes (:addr-recv payload) true)
      (encode-version-cmd>=106 version payload)
;      (encode-version-cmd>=70001 version payload)
      )))
                  
  

(defn open-socket [host port]
  (let [s (Socket. host port)]
    {:socket s
     :in (BufferedInputStream. (.getInputStream s))
     :out (BufferedOutputStream. (.getOutputStream s))}
    ))

(def mainnet 0xD9B4BEF9)
(def testnet3 0x0709110B)


(defn send-ver-msg [s ver-msg]
  (.write (:out s) (byte-array (msg->bytes ver-msg mainnet)))
  (.flush (:out s))
  (let [ba (byte-array 200)]
    (.read (:in s) ba)
    (dbg (map #(Integer/toString % 16) (bytes->unsigned ba)))
    (read-msg (ByteArrayInputStream. ba) mainnet)))

(def a-ver-msg 
  {:cmd "version", :payload 
   {:version 60002, 
    :timestamp (.getEpochSecond (Instant/parse "2012-12-18T02:12:33.00Z")) 
    :services {},
    :nounce (long (rand Long/MAX_VALUE))
    :user-agent "Satoshi:0.7.2"
    :start-height 212672
    :relay 0
    :addr-from
    {:time 0, :ip [88 129 173 173], 
     :port 8333}
    :addr-recv
    {:time 0, :ip [103 80 168 57], 
     :port 8333}}})



   
                            